src/Abp.ZeroCore.EntityFrameworkCore/EntityHistory/EntitySnapshotManager.cs

```diff
@@ -0,0 +1,93 @@
+        {
+            return await GetSnapshotAsync<TEntity, int>(id, snapshotTime);
+        }
+        public async Task<EntityHistorySnapshot> GetSnapshotAsync<TEntity, TPrimaryKey>(TPrimaryKey id, DateTime snapshotTime) where TEntity : class, IEntity<TPrimaryKey>
+        {
+            var entity = _entityChangeRepository.GetDbContext()
+                .Set<TEntity>().AsQueryable().FirstOrDefault(x => x.Id.Equals(id));
+
+            var snapshotPropertiesDictionary = new Dictionary<string, string>();
+            var propertyChangesStackTreeDictionary = new Dictionary<string, string>();
+
+            if (entity != null)
+            {
+                string fullName = typeof(TEntity).FullName;
+                var idJson = id.ToJsonString();
+
+                var allChanges = _entityChangeRepository.GetAll().Include(x => x.PropertyChanges).ToList();
+                var changes = await _entityChangeRepository.GetAll()//select all changes which created after snapshot time 
+                    .Where(x => x.EntityTypeFullName == fullName && x.EntityId == idJson && x.ChangeTime > snapshotTime && x.ChangeType != EntityChangeType.Created)
+                    .OrderByDescending(x => x.ChangeTime)
+                    .Select(x => new { x.ChangeType, x.PropertyChanges }).ToListAsync();
+
+                //and revoke all changes
+
+                foreach (var change in changes)// desc ordered changes
+                {
+                    foreach (var entityPropertyChange in change.PropertyChanges)
+                    {
+                        if (snapshotPropertiesDictionary.ContainsKey(entityPropertyChange.PropertyName))
+                        {
+                            snapshotPropertiesDictionary[entityPropertyChange.PropertyName] = entityPropertyChange.OriginalValue;//set back to orginal value
+                        }
+                        else
+                        {
+                            snapshotPropertiesDictionary.Add(entityPropertyChange.PropertyName, entityPropertyChange.OriginalValue);//set back to orginal value
+                        }
+
+                        //create change stack tree
+                        if (propertyChangesStackTreeDictionary.ContainsKey(entityPropertyChange.PropertyName))
+                        {
+                            propertyChangesStackTreeDictionary[entityPropertyChange.PropertyName] += " -> " + entityPropertyChange.OriginalValue;
+                        }
+                        else
+                        {
+                            string propertyCurrentValue = "PropertyNotExist";
+
+                            var propertyInfo = typeof(TEntity).GetProperty(entityPropertyChange.PropertyName);
+                            if (propertyInfo != null)
+                            {
+                                var val = propertyInfo.GetValue(entity);
```

Comment:
how do we define current value? why not using `NewValue` from `PropertyChange`

Reply:
To check whether the property exists in new entity. Property of entity might be deleted after that recorded history.